// realm-object-server/dist/realms/GlobalManagementRealm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class PermissionChange {
}
PermissionChange.schema = {
    name: 'PermissionChange',
    primaryKey: 'id',
    properties: {
        id: { type: 'string' },
        createdAt: { type: 'date', default: new Date() },
        updatedAt: { type: 'date', default: new Date() },
        statusCode: { type: 'int', optional: true },
        statusMessage: { type: 'string', optional: true },
        userId: { type: 'string' },
        realmUrl: { type: 'string' },
        metadataKey: { type: 'string', optional: true },
        metadataValue: { type: 'string', optional: true },
        mayRead: { type: 'bool', optional: true },
        mayWrite: { type: 'bool', optional: true },
        mayManage: { type: 'bool', optional: true },
    }
};
exports.PermissionChange = PermissionChange;
class PermissionOffer {
}
PermissionOffer.schema = {
    name: 'PermissionOffer',
    primaryKey: 'id',
    properties: {
        id: { type: 'string', optional: false, indexed: true },
        createdAt: { type: 'date', default: new Date() },
        updatedAt: { type: 'date', default: new Date() },
        statusCode: { type: 'int', optional: true },
        statusMessage: { type: 'string', optional: true },
        token: { type: 'string', optional: true, indexed: true },
        realmUrl: { type: 'string' },
        mayRead: { type: 'bool', default: false },
        mayWrite: { type: 'bool', default: false },
        mayManage: { type: 'bool', default: false },
        expiresAt: { type: 'date', optional: true }
    }
};
exports.PermissionOffer = PermissionOffer;
class PermissionOfferResponse {
}
PermissionOfferResponse.schema = {
    name: 'PermissionOfferResponse',
    primaryKey: 'id',
    properties: {
        id: { type: 'string', optional: false },
        createdAt: { type: 'date', default: new Date() },
        updatedAt: { type: 'date', default: new Date() },
        statusCode: { type: 'int', optional: true },
        statusMessage: { type: 'string', optional: true },
        token: { type: 'string' },
        realmUrl: { type: 'string', optional: true }
    }
};
exports.PermissionOfferResponse = PermissionOfferResponse;
