// realm-object-server/dist/termsofservice.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Mixpanel = require("mixpanel");
const fs = require("fs-extra");
const path = require("path");
const promptly = require("promptly");
const mixpanel = Mixpanel.init('953cbc7bc5e8daa02b91425ac92d36dd');
function writeAgreementJson(path, email, terms, agreeToLearnAboutUpdatesAndFixes) {
    return __awaiter(this, void 0, void 0, function* () {
        const json = {
            email: email,
            terms: terms,
            agreeToLearnAboutUpdatesAndFixes: agreeToLearnAboutUpdatesAndFixes
        };
        yield fs.writeFile(path, JSON.stringify(json, null, 4), { encoding: 'utf8' });
    });
}
function attemptToSubmitToMixpanel(email, agreeToLearnAboutUpdatesAndFixes) {
    return __awaiter(this, void 0, void 0, function* () {
        mixpanel.track('ros_agreed_to_terms', {
            version: (yield fs.readJson(path.join(__dirname, '../package.json'))).version,
            email: email,
            agreeToLearnAboutUpdatesAndFixes: agreeToLearnAboutUpdatesAndFixes
        });
    });
}
function validateEmail(email) {
    const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    const passesTest = re.test(email);
    if (!passesTest) {
        throw new Error('Please enter a valid email');
    }
    return email;
}
function agreeToTerms() {
    return __awaiter(this, void 0, void 0, function* () {
        if (process.env.ROS_SKIP_PROMPTS) {
            return;
        }
        const terms = 'https://realm.io/legal/developer-license-terms/';
        const tosPath = process.env.DOCKER_DATA_PATH ?
            path.resolve(process.env.DOCKER_DATA_PATH, "realm-object-server-agreement.json") :
            path.resolve(__dirname, '../realm-object-server-agreement.json');
        const jsonExists = yield fs.pathExists(tosPath);
        if (jsonExists) {
            const agreementFile = yield fs.readFile(tosPath);
            const agreement = JSON.parse(agreementFile.toString());
            const exampleAgreement = {
                email: 'realm@realm.io',
                terms: terms,
                agreeToLearnAboutUpdatesAndFixes: true,
            };
            const requiredKeys = ['email', 'terms', 'agreeToLearnAboutUpdatesAndFixes'];
            for (var key of requiredKeys) {
                if (!(key in agreement)) {
                    throw new Error("Invalid format of 'realm-object-server-agreement.json'. Expected something like:\n " +
                        JSON.stringify(exampleAgreement, null, 4) + "\nPlease delete that file and try again.");
                }
            }
            const agreementEmail = agreement['email'];
            try {
                validateEmail(agreementEmail);
            }
            catch (Error) {
                throw new Error("Invalid email format in 'realm-object-server-agreement.json': " + agreementEmail);
            }
            if (agreement['terms'] !== terms) {
                throw new Error("Invalid terms URL. Must be: " + terms);
            }
            return;
        }
        let email;
        let agreeToLearnAboutUpdatesAndFixes = false;
        if (process.env.DOCKER_DATA_PATH) {
            email = validateEmail(process.env.ROS_TOS_EMAIL_ADDRESS);
        }
        else {
            console.log('Please agree to our terms of service by siging with your email. You can read the full terms at: ' + terms);
            email = yield promptly.prompt(`Agree with your email: `, { validator: validateEmail });
            agreeToLearnAboutUpdatesAndFixes = yield promptly.confirm('Would you like to receive product updates and updates about critical bug fixes? (Y/N)');
        }
        attemptToSubmitToMixpanel(email, agreeToLearnAboutUpdatesAndFixes).catch((err) => { });
        yield writeAgreementJson(tosPath, email, terms, agreeToLearnAboutUpdatesAndFixes);
    });
}
exports.agreeToTerms = agreeToTerms;
